File: chron/services/scheduler/TriggerService.ts
// TriggerService.ts 

import type { IPlatform } from "../platform/IPlatform";
import type { ILogger } from "../logging/ILogger";
import type { ICallActions } from "../call/ICallActions";
import { EventManager } from "../events/events";


export class TriggerService {
  private readonly platform: IPlatform;
  private readonly logger: ILogger;
  private readonly events: EventManager;
  private readonly call: ICallActions;

  constructor(
    platform: IPlatform,
    logger: ILogger,
    events: EventManager,
    call: ICallActions
  ) {
    this.platform = platform;
    this.logger = logger;
    this.events = events;
    this.call = call;
  }

  // Called on each alarm tick
  async runTrigger(): Promise<void> {
    // requires vault setup
    if (!(await this.platform.vaultIsInitialized())) {
      this.logger.info("[Trigger] Vault not initialized -> skip");
      return;
    }

    // bypass battery + DND checks, no notification
    const activeEvent = await this.events.getActiveEventName();
    if (activeEvent !== null) {
      this.logger.info("[Trigger] Event override -> auto-log: " + activeEvent);
      this.call.autoLogEvent(activeEvent);
      return;
    }

    // Battery Level: skip if < 5%
    if (!(await this.platform.isBatterySafe())) {
      this.logger.info("[Trigger] Battery too low -> skip");
      return;
    }

    // skip if any DND mode
    if (!(await this.platform.isDndDisabled())) {
      this.logger.info("[Trigger] DND active -> skip");
      return;
    }

    // start call experience
    this.logger.info("[Trigger] Starting call interaction");
    this.call.startIncomingCall();
  }
}

File: legacy-app/core/scheduler/TriggerService.cpp
#include "scheduler/TriggerService.hpp"

TriggerService::TriggerService(IPlatform& platform, ILogger& logger, EventManager& events, ICallActions& call)
    : m_platform(platform), m_logger(logger), m_events(events), m_call(call)
{
}

void TriggerService::runTrigger()
{
    // requires vault setup
    if (!m_platform.vaultIsInitialized())
    {
        m_logger.info("[Trigger] Vault not initialized -> skip");
        return;
    }

    // bypass battery + DND checks, no notification
    if (auto activeEvent = m_events.getActiveEventName(); activeEvent.has_value())
    {
        m_logger.info("[Trigger] Event override -> auto-log: " + activeEvent.value());
        m_call.autoLogEvent(activeEvent.value());
        return;
    }

    // Battery Level: skip if < 5%
    if (!m_platform.isBatterySafe())
    {
        m_logger.info("[Trigger] Battery too low -> skip");
        return;
    }

    // skip if any DND mode
    if (!m_platform.isDndDisabled())
    {
        m_logger.info("[Trigger] DND active -> skip");
        return;
    }

    // start call experience
    m_logger.info("[Trigger] Starting call interaction");
    m_call.startIncomingCall();
}